cluster-autoscaler/core/utils.go

```diff
@@ -173,34 +178,38 @@ func FilterOutExpendablePods(pods []*apiv1.Pod, expendablePodsPriorityCutoff int
 	return result
 }
 
-// FilterSchedulablePodsForNode filters pods that can be scheduled on the given node.
-func FilterSchedulablePodsForNode(context *context.AutoscalingContext, pods []*apiv1.Pod, nodeGroupId string, nodeInfo *schedulercache.NodeInfo) []*apiv1.Pod {
-	schedulablePods := []*apiv1.Pod{}
+// CheckPodsSchedulableOnNode checks if pods can be scheduled on the given node.
+func CheckPodsSchedulableOnNode(context *context.AutoscalingContext, pods []*apiv1.Pod, nodeGroupId string, nodeInfo *schedulercache.NodeInfo) map[*apiv1.Pod]*simulator.PredicateError {
+	schedulingErrors := map[*apiv1.Pod]*simulator.PredicateError{}
 	loggingQuota := glogx.PodsLoggingQuota()
 	podSchedulable := make(podSchedulableMap)
 	for _, pod := range pods {
-		schedulable, found := podSchedulable.get(pod)
-		if found {
-			if schedulable {
-				schedulablePods = append(schedulablePods, pod)
-			} else {
+		schedulable, err, found := podSchedulable.get(pod)
+		if !found {
+			err = context.PredicateChecker.CheckPredicates(pod, nil, nodeInfo)
+			schedulable = err == nil
+			podSchedulable.set(pod, schedulable, err)
+		}
+
+		// Check if pod isn't repeated before overwriting result for it.
+		if _, repeated := schedulingErrors[pod]; repeated {
+			// This shouldn't really happen.
+			glog.Warningf("Pod %v appears multiple time on pods list, will only count it once in scale-up simulation", pod)
+		}
+
+		schedulingErrors[pod] = err
+
+		if !schedulable {
+			if found {
```

Comment:
Similar to previous method. I'd expect found to only show up at the beginning of the method. As soon as we need to do anything custom for cached result I find old flow much easier to follow.

Reply:
Changed back to checking schedulable in each branch, PTAL